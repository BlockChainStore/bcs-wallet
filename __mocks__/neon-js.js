const neonjs = jest.genMockFromModule('neon-js')

const promiseMockGen = (result, error = false) => {
  return jest.fn(() => {
    return new Promise((resolve, reject) => {
      if (error) reject(error)
      resolve(result)
    })
  })
}

const privateKey = 'L4AJ14CNaBWPemRJKC34wyZwbmxg33GETs4Y1F8uK7rRmZ2UHrJn'
const address = 'AM22coFfbe9N6omgL9ucFBLkeaMNg9TEyL'
const encryptedKey = '6PYUGtvXiT5TBetgWf77QyAFidQj61V8FJeFBFtYttmsSxcbmP4vCFRCWu'
const scriptHash = '4bcdc110b6514312ead9420467475232d4f08539'

neonjs.api = {
  neonDB: {
    getClaims: jest.fn(),
    doClaimAllGas: promiseMockGen({ result: true }),
    doSendAsset: promiseMockGen({ result: true }),
    getWalletDBHeight: promiseMockGen(586435),
    getAPIEndpoint: jest.fn(() => 'http://testnet-api.wallet.cityofzion.io'),
    getRPCEndpoint: promiseMockGen(''),
    doMintTokens: promiseMockGen({ result: true }),
    getTransactionHistory: promiseMockGen([]),
    getBalance: promiseMockGen({ NEO: { balance: 1 }, GAS: { balance: 1 } })
  },
  nep5: {
    getTokenInfo: promiseMockGen({ result: true }),
    getTokenBalance: promiseMockGen(100)
  }
}

neonjs.create = {
  account: {
    address
    // privateKey
    // wif
  }
}
neonjs.tx = {
  serializeTransaction: jest.fn()
}
// TODO - look into why I chose to use encrypt vs encryptWif from new API
neonjs.wallet = {
  getPublicKeyEncoded: jest.fn(),
  decryptWIF: jest.fn((wif) => {
    return new Promise((resolve, reject) => {
      if (!wif) reject(new Error())
      resolve(privateKey)
    })
  }),
  decrypt: jest.fn(() => privateKey),
  encrypt: jest.fn(() => encryptedKey),
  generatePrivateKey: jest.fn(() => privateKey),
  Account: jest.fn(() => { return { address } }),
  getVerificationScriptFromPublicKey: jest.fn(() => scriptHash),
  isAddress: jest.fn(() => true),
  isNEP2: jest.fn(() => true)
}

module.exports = neonjs
