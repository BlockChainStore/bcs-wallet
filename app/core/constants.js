// @flow
export const NEON_WALLET_RELEASE_LINK =
  'https://github.com/CityOfZion/neon-wallet/releases'

export const ASSETS = {
  NEO: 'NEO',
  GAS: 'GAS'
}

export const EXPLORERS = {
  NEO_TRACKER: 'Neotracker',
  NEO_SCAN: 'Neoscan',
  ANT_CHAIN: 'Antchain',
  NEO_VERSE: 'NEOverse'
}

export const ROUTES = {
  HOME: '',
  DASHBOARD: '/dashboard',
  CREATE_WALLET: '/create',
  ENCRYPT_KEY: '/encrypt-key',
  LOGIN_PRIVATE_KEY: '/login-private-key',
  LOGIN_LOCAL_STORAGE: '/login-local-storage',
  LOGIN_LEDGER_NANO_S: '/login-ledger-nano-s',
  LOGIN_NEP2: '/login-encrypted',
  LOGIN_TOKEN_SALE: '/login-token-sale',
  TOKEN_SALE: '/token-sale',
  SETTINGS: '/settings',
  DISPLAY_WALLET_KEYS: '/display-wallet-keys'
}

export const NOTIFICATION_LEVELS = {
  ERROR: 'error',
  SUCCESS: 'success',
  INFO: 'info',
  WARNING: 'warning'
}

export const NOTIFICATION_POSITIONS = {
  TOP_CENTER: 'tc',
  TOP_RIGHT: 'tr',
  TOP_LEFT: 'tl',
  BOTTOM_CENTER: 'bc',
  BOTTOM_RIGHT: 'br',
  BOTTOM_LEFT: 'bl'
}

export const BIP44_PATH =
  '8000002C' + '80000378' + '80000000' + '00000000' + '00000000'

export const MODAL_TYPES = {
  SEND: 'SEND',
  RECEIVE: 'RECEIVE',
  CONFIRM: 'CONFIRM',
  TOKEN_INFO: 'TOKEN_INFO',
  TOKEN: 'TOKEN',
  ICO: 'ICO'
}

export const MAIN_NETWORK_ID = '1'
export const TEST_NETWORK_ID = '2'

// TestNet
export const TOKENS_TEST = {
  BCS : 'e81c837d37763afbf894d319c868a09f219f2be9'
}

// MainNet
export const TOKENS = {
  BCS : '546c5872a992b2754ef327154f4c119baabff65f'
}

export const DEFAULT_WALLET = {
  name: 'userWallet',
  version: '1.0',
  scrypt: {
    cost: 16384,
    blockSize: 8,
    parallel: 8,
    size: 64
  },
  accounts: [],
  extra: null
}

export const DEFAULT_CURRENCY_CODE = 'usd'

export const CURRENCIES = {
  aud: { symbol: '$' },
  brl: { symbol: 'R$' },
  cad: { symbol: '$' },
  chf: { symbol: 'Fr.' },
  clp: { symbol: '$' },
  cny: { symbol: '¥' },
  czk: { symbol: 'Kč' },
  dkk: { symbol: 'kr. ' },
  eur: { symbol: '€' },
  gbp: { symbol: '£' },
  hkd: { symbol: '$' },
  huf: { symbol: 'Ft ' },
  idr: { symbol: 'Rp ' },
  ils: { symbol: '₪' },
  inr: { symbol: '₹' },
  jpy: { symbol: '¥' },
  krw: { symbol: '₩' },
  mxn: { symbol: '$' },
  myr: { symbol: 'RM' },
  nok: { symbol: 'kr ' },
  nzd: { symbol: '$' },
  php: { symbol: '₱' },
  pkr: { symbol: '₨ ' },
  pln: { symbol: 'zł' },
  rub: { symbol: '₽' },
  sek: { symbol: 'kr ' },
  sgd: { symbol: 'S$' },
  thb: { symbol: '฿' },
  try: { symbol: '₺' },
  twd: { symbol: 'NT$' },
  usd: { symbol: '$' },
  zar: { symbol: 'R ' }
}
