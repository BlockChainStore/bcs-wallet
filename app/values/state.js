// @flow
export type ProgressState = 'INITIAL' | 'LOADING' | 'LOADED' | 'FAILED'

export const INITIAL: ProgressState = 'INITIAL'
export const LOADING: ProgressState = 'LOADING'
export const LOADED: ProgressState = 'LOADED'
export const FAILED: ProgressState = 'FAILED'
