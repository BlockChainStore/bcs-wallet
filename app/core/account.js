// @flow
import { wallet } from 'neon-js'

import { getStorage, setStorage } from './storage'

function getNEP6AddressData (matchesEncryptedWIF: boolean, address: string) {
  if (matchesEncryptedWIF) {
    return { address }
  } else {
    return {}
  }
}

export async function upgradeNEP6AddAddresses (encryptedWIF: string, wif: string) {
  const data = getStorage('userWallet')
  const loggedIntoAccount = new wallet.Account(wif)

  if (data && data.accounts) {
    const accounts = data.accounts.map((account, idx) => ({
      ...account,
      ...getNEP6AddressData(account.key === encryptedWIF, loggedIntoAccount.address)
    }))

    await setStorage('userWallet', { ...data, accounts })
  }
}
