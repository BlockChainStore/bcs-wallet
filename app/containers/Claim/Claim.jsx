// @flow
import React, { Component } from 'react'

import Button from '../../components/Button'
import Tooltip from '../../components/Tooltip'
import { formatGAS } from '../../core/formatters'
import { toBigNumber } from '../../core/math'

type Props = {
  doClaimNotify: Function,
  setClaimRequest: Function,
  doGasClaim: Function,
  claimRequest: boolean,
  disableClaimButton: boolean,
  claimAmount: string,
}

export default class Claim extends Component<Props> {
  componentDidUpdate (prevProps: Props) {
    const { claimRequest, doClaimNotify, setClaimRequest } = this.props
    if (claimRequest && !prevProps.claimRequest) {
      setClaimRequest(false)
      doClaimNotify()
    }
  }

  render () {
    const { claimAmount, disableClaimButton, doGasClaim } = this.props
    const shouldDisableButton = disableClaimButton || toBigNumber(claimAmount).eq(0)
    const formattedAmount = formatGAS(claimAmount)
    return (
      <div>
        <Tooltip
          title='You can claim GAS once every 5 minutes'
          disabled={!disableClaimButton}
        >
          <Button id='claim' disabled={shouldDisableButton} onClick={doGasClaim}>
            Claim {formattedAmount} GAS
          </Button>
        </Tooltip>
      </div>
    )
  }
}
