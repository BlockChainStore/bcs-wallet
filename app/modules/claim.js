// @flow
import { api } from 'neon-js'

import {
  showErrorNotification,
  showSuccessNotification,
  showInfoNotification
} from './notifications'
import {
  getNetwork,
  getWIF,
  getAddress,
  getSigningFunction,
  getPublicKey,
  getIsHardwareLogin,
  getNEO
} from '../core/deprecated'
import { ASSETS } from '../core/constants'
import asyncWrap from '../core/asyncHelper'
import { FIVE_MINUTES_MS } from '../core/time'

import { log } from '../util/Logs'
import { toNumber } from '../core/math'

// Constants
export const SET_CLAIM_REQUEST = 'SET_CLAIM_REQUEST'
export const DISABLE_CLAIM = 'DISABLE_CLAIM'

// Actions
export function setClaimRequest (claimRequest: boolean) {
  return {
    type: SET_CLAIM_REQUEST,
    payload: { claimRequest }
  }
}

export function disableClaim (disableClaimButton: boolean) {
  return {
    type: DISABLE_CLAIM,
    payload: { disableClaimButton }
  }
}

export const doClaimNotify = () => async (
  dispatch: DispatchType,
  getState: GetStateType
) => {
  const state = getState()
  const wif = getWIF(state)
  const address = getAddress(state)
  const net = getNetwork(state)
  const signingFunction = getSigningFunction(state)
  const publicKey = getPublicKey(state)
  const isHardwareClaim = getIsHardwareLogin(state)

  log(net, 'CLAIM', address, { info: 'claim all GAS' })

  let claimGasFn
  if (isHardwareClaim) {
    dispatch(
      showInfoNotification({
        message:
          'Sign transaction 2 of 2 to claim GAS on your hardware device (claiming GAS)',
        autoDismiss: 0
      })
    )
    claimGasFn = () =>
      api.neonDB.doClaimAllGas(net, publicKey, signingFunction)
  } else {
    claimGasFn = () => api.neonDB.doClaimAllGas(net, wif, null)
  }

  const [err, response] = await asyncWrap(claimGasFn())
  if (!err && response.result) {
    dispatch(
      showSuccessNotification({
        message:
          'Claim was successful! Your balance will update once the blockchain has processed it.'
      })
    )
    setTimeout(() => dispatch(disableClaim(false)), FIVE_MINUTES_MS)
  } else {
    return dispatch(showErrorNotification({ message: 'Claim failed' }))
  }
}

// To initiate claim, first send all NEO to own address, the set claimRequest state
// When new claims are available, this will trigger the claim
export const doGasClaim = () => async (
  dispatch: DispatchType,
  getState: GetStateType
) => {
  const state = getState()
  const wif = getWIF(state)
  const address = getAddress(state)
  const net = getNetwork(state)
  const NEO = getNEO(state)
  const signingFunction = getSigningFunction(state)
  const publicKey = getPublicKey(state)
  const isHardwareClaim = getIsHardwareLogin(state)

  // if no NEO in account, no need to send to self first
  if (NEO === '0') {
    return dispatch(doClaimNotify())
  } else {
    dispatch(
      showInfoNotification({
        message: 'Sending NEO to Yourself...',
        autoDismiss: 0
      })
    )
    log(net, 'SEND', address, { to: address, amount: NEO, asset: ASSETS.NEO })

    let sendAssetFn
    if (isHardwareClaim) {
      dispatch(
        showInfoNotification({
          message:
            'Sign transaction 1 of 2 to claim GAS on your hardware device (sending NEO to yourself)',
          autoDismiss: 0
        })
      )
      sendAssetFn = () =>
        api.neonDB.doSendAsset(
          net,
          address,
          publicKey,
          { [ASSETS.NEO]: toNumber(NEO) },
          signingFunction
        )
    } else {
      sendAssetFn = () =>
        api.neonDB.doSendAsset(net, address, wif, { [ASSETS.NEO]: toNumber(NEO) }, null)
    }

    const [err, response] = await asyncWrap(sendAssetFn())
    if (err || response.result === undefined || response.result === false) {
      return dispatch(
        showErrorNotification({ message: 'Transaction failed!' })
      )
    } else {
      dispatch(
        showInfoNotification({
          message: 'Waiting for transaction to clear...',
          autoDismiss: 0
        })
      )
      dispatch(setClaimRequest(true))
      return dispatch(disableClaim(true))
    }
  }
}

// State Getters
export const getClaimRequest = (state: Object) => state.claim.claimRequest
export const getDisableClaimButton = (state: Object) => state.claim.disableClaimButton

const initialState = {
  claimRequest: false,
  disableClaimButton: false
}

export default (state: Object = initialState, action: ReduxAction) => {
  switch (action.type) {
    case SET_CLAIM_REQUEST:
      const { claimRequest } = action.payload
      return {
        ...state,
        claimRequest
      }
    case DISABLE_CLAIM:
      const { disableClaimButton } = action.payload
      return {
        ...state,
        disableClaimButton
      }
    default:
      return state
  }
}
