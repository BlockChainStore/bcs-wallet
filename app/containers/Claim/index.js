// @flow
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { compose } from 'recompose'

import Claim from './Claim'
import claimsActions from '../../actions/claimsActions'
import withData from '../../hocs/api/withData'
import {
  setClaimRequest,
  doGasClaim,
  doClaimNotify,
  getClaimRequest,
  getDisableClaimButton
} from '../../modules/claim'

const mapStateToProps = (state: Object) => ({
  claimRequest: getClaimRequest(state),
  disableClaimButton: getDisableClaimButton(state)
})

const actionCreators = {
  setClaimRequest,
  doGasClaim,
  doClaimNotify
}

const mapDispatchToProps = dispatch => bindActionCreators(actionCreators, dispatch)

const mapClaimsDataToProps = (claims) => ({
  claimAmount: claims.total
})

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withData(claimsActions, mapClaimsDataToProps)
)(Claim)
