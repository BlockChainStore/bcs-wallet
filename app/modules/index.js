// @flow
import { combineReducers } from 'redux'
import api from './api'
import generateWallet from './generateWallet'
import claim from './claim'
import notifications from './notifications'
import modal from './modal'
import addressBook from './addressBook'

export default combineReducers({
  api,
  generateWallet,
  claim,
  notifications,
  modal,
  addressBook
})
