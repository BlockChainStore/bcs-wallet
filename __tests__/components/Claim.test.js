import React from 'react'
import * as neonjs from 'neon-js'
import { merge } from 'lodash'
import { mount } from 'enzyme'

import { createStore, provideStore } from '../testHelpers'
import Claim from '../../app/containers/Claim'
import { setClaimRequest, disableClaim } from '../../app/modules/claim'
import { SHOW_NOTIFICATION } from '../../app/modules/notifications'
import { NOTIFICATION_LEVELS, MAIN_NETWORK_ID } from '../../app/core/constants'
import { LOADED } from '../../app/values/state'

jest.useFakeTimers()

const initialState = {
  api: {
    NETWORK: {
      batch: false,
      state: LOADED,
      data: MAIN_NETWORK_ID
    },
    CLAIMS: {
      batch: false,
      state: LOADED,
      data: {
        total: '0.01490723'
      }
    },
    BALANCES: {
      batch: false,
      state: LOADED,
      data: {
        NEO: '10',
        GAS: '10.00000000'
      }
    }
  },
  claim: {
    claimRequest: false,
    disableClaimButton: false
  }
}

const simulateSendAsset = (result) => {
  return jest.fn(() => {
    return new Promise((resolve, reject) => {
      resolve({ result })
    })
  })
}

describe('Claim', () => {
  test('should render without crashing', () => {
    const store = createStore(initialState)
    const wrapper = mount(provideStore(<Claim />, store))
    expect(wrapper).toMatchSnapshot()
  })

  test('should render claim GAS button as enabled', () => {
    const store = createStore(initialState)
    const wrapper = mount(provideStore(<Claim />, store))
    expect(wrapper).toMatchSnapshot()
  })

  test('should render claim GAS button as disabled', () => {
    const state = merge({}, initialState, { claim: { disableClaimButton: true } })
    const store = createStore(state)
    const wrapper = mount(provideStore(<Claim />, store))
    expect(wrapper).toMatchSnapshot()
  })

  describe('when claim GAS button is clicked', () => {
    test('should dispatch transaction failure event', async (done) => {
      const store = createStore(initialState)
      const wrapper = mount(provideStore(<Claim />, store))
      neonjs.api.neonDB.doSendAsset = simulateSendAsset(false)

      wrapper.find('button#claim').simulate('click')
      jest.runAllTimers()
      await Promise.resolve().then().then().then()

      const actions = store.getActions()

      expect(actions).toContainEqual(expect.objectContaining({
        type: SHOW_NOTIFICATION,
        payload: expect.objectContaining({
          level: NOTIFICATION_LEVELS.ERROR,
          message: 'Transaction failed!'
        })
      }))
      done()
    })

    test('should dispatch transaction waiting, set claim request, and disable claim events', async (done) => {
      const store = createStore(initialState)
      const wrapper = mount(provideStore(<Claim />, store))
      neonjs.api.neonDB.doSendAsset = simulateSendAsset(true)

      wrapper.find('button#claim').simulate('click')
      jest.runAllTimers()
      await Promise.resolve().then().then().then()

      const actions = store.getActions()

      expect(actions).toContainEqual(setClaimRequest(true))
      expect(actions).toContainEqual(disableClaim(true))
      done()
    })
  })
})
