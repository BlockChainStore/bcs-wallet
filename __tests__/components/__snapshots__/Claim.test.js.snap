// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Claim should render claim GAS button as disabled 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Connect(withData(Claim)))>
    <Connect(withData(Claim))
      claimRequest={false}
      disableClaimButton={true}
      doClaimNotify={[Function]}
      doGasClaim={[Function]}
      setClaimRequest={[Function]}
    >
      <withData(Claim)
        claimAmount="0.01490723"
        claimRequest={false}
        disableClaimButton={true}
        dispatch={[Function]}
        doClaimNotify={[Function]}
        doGasClaim={[Function]}
        setClaimRequest={[Function]}
      >
        <div>
          <Tooltip
            disabled={false}
            title="You can claim GAS once every 5 minutes"
          >
            <Tooltip
              animateFill={true}
              animation="fade"
              arrow={true}
              arrowSize="regular"
              className=""
              delay={0}
              disabled={false}
              distance={10}
              duration={375}
              followCursor={false}
              hideDelay={0}
              hideDuration={375}
              hideOnClick={true}
              html={null}
              inertia={false}
              interactive={false}
              interactiveBorder={2}
              multiple={false}
              offset={0}
              onHidden={[Function]}
              onHide={[Function]}
              onRequestClose={[Function]}
              onShow={[Function]}
              onShown={[Function]}
              popperOptions={Object {}}
              position="bottom"
              size="regular"
              sticky={false}
              stickyDuration={200}
              style={Object {}}
              theme="dark"
              title="You can claim GAS once every 5 minutes"
              touchHold={false}
              trigger="mouseenter focus"
              unmountHTMLWhenHide={false}
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline",
                  }
                }
                title="You can claim GAS once every 5 minutes"
              >
                <Button
                  cancel={false}
                  disabled={true}
                  id="claim"
                  onClick={[Function]}
                  primary={true}
                  secondary={false}
                >
                  <button
                    className="button primary"
                    disabled={true}
                    id="claim"
                    onClick={[Function]}
                    type="button"
                  >
                    Claim 
                    0.01490723
                     GAS
                  </button>
                </Button>
              </div>
            </Tooltip>
          </Tooltip>
        </div>
      </withData(Claim)>
    </Connect(withData(Claim))>
  </Connect(Connect(withData(Claim)))>
</Provider>
`;

exports[`Claim should render claim GAS button as enabled 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Connect(withData(Claim)))>
    <Connect(withData(Claim))
      claimRequest={false}
      disableClaimButton={false}
      doClaimNotify={[Function]}
      doGasClaim={[Function]}
      setClaimRequest={[Function]}
    >
      <withData(Claim)
        claimAmount="0.01490723"
        claimRequest={false}
        disableClaimButton={false}
        dispatch={[Function]}
        doClaimNotify={[Function]}
        doGasClaim={[Function]}
        setClaimRequest={[Function]}
      >
        <div>
          <Tooltip
            disabled={true}
            title="You can claim GAS once every 5 minutes"
          >
            <Tooltip
              animateFill={true}
              animation="fade"
              arrow={true}
              arrowSize="regular"
              className=""
              delay={0}
              disabled={true}
              distance={10}
              duration={375}
              followCursor={false}
              hideDelay={0}
              hideDuration={375}
              hideOnClick={true}
              html={null}
              inertia={false}
              interactive={false}
              interactiveBorder={2}
              multiple={false}
              offset={0}
              onHidden={[Function]}
              onHide={[Function]}
              onRequestClose={[Function]}
              onShow={[Function]}
              onShown={[Function]}
              popperOptions={Object {}}
              position="bottom"
              size="regular"
              sticky={false}
              stickyDuration={200}
              style={Object {}}
              theme="dark"
              title="You can claim GAS once every 5 minutes"
              touchHold={false}
              trigger="mouseenter focus"
              unmountHTMLWhenHide={false}
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline",
                  }
                }
                title="You can claim GAS once every 5 minutes"
              >
                <Button
                  cancel={false}
                  disabled={false}
                  id="claim"
                  onClick={[Function]}
                  primary={true}
                  secondary={false}
                >
                  <button
                    className="button primary"
                    disabled={false}
                    id="claim"
                    onClick={[Function]}
                    type="button"
                  >
                    Claim 
                    0.01490723
                     GAS
                  </button>
                </Button>
              </div>
            </Tooltip>
          </Tooltip>
        </div>
      </withData(Claim)>
    </Connect(withData(Claim))>
  </Connect(Connect(withData(Claim)))>
</Provider>
`;

exports[`Claim should render without crashing 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Connect(withData(Claim)))>
    <Connect(withData(Claim))
      claimRequest={false}
      disableClaimButton={false}
      doClaimNotify={[Function]}
      doGasClaim={[Function]}
      setClaimRequest={[Function]}
    >
      <withData(Claim)
        claimAmount="0.01490723"
        claimRequest={false}
        disableClaimButton={false}
        dispatch={[Function]}
        doClaimNotify={[Function]}
        doGasClaim={[Function]}
        setClaimRequest={[Function]}
      >
        <div>
          <Tooltip
            disabled={true}
            title="You can claim GAS once every 5 minutes"
          >
            <Tooltip
              animateFill={true}
              animation="fade"
              arrow={true}
              arrowSize="regular"
              className=""
              delay={0}
              disabled={true}
              distance={10}
              duration={375}
              followCursor={false}
              hideDelay={0}
              hideDuration={375}
              hideOnClick={true}
              html={null}
              inertia={false}
              interactive={false}
              interactiveBorder={2}
              multiple={false}
              offset={0}
              onHidden={[Function]}
              onHide={[Function]}
              onRequestClose={[Function]}
              onShow={[Function]}
              onShown={[Function]}
              popperOptions={Object {}}
              position="bottom"
              size="regular"
              sticky={false}
              stickyDuration={200}
              style={Object {}}
              theme="dark"
              title="You can claim GAS once every 5 minutes"
              touchHold={false}
              trigger="mouseenter focus"
              unmountHTMLWhenHide={false}
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline",
                  }
                }
                title="You can claim GAS once every 5 minutes"
              >
                <Button
                  cancel={false}
                  disabled={false}
                  id="claim"
                  onClick={[Function]}
                  primary={true}
                  secondary={false}
                >
                  <button
                    className="button primary"
                    disabled={false}
                    id="claim"
                    onClick={[Function]}
                    type="button"
                  >
                    Claim 
                    0.01490723
                     GAS
                  </button>
                </Button>
              </div>
            </Tooltip>
          </Tooltip>
        </div>
      </withData(Claim)>
    </Connect(withData(Claim))>
  </Connect(Connect(withData(Claim)))>
</Provider>
`;
